### PR TITLE
Move auth from user module

### PR DESCRIPTION
**Pull request by basilalex**
This PR moves auth logic to a separate module.

### CODE REVIEW

File: modules/user/client-react/auth/google/containers/GoogleButton.native.jsx [hunk 1]
```diff
@@ -14,9 +14,9 @@ import {
   btnTextContainer,
   btnText
 } from '@module/look-client-react-native/styles';
+import access from '@module/authentication-client-react';
```
Comment (evgenijolshanskij) [inline modules/user/client-react/auth/google/containers/GoogleButton.native.jsx]: All of those components shouldn't be in the User module.

File: modules/authentication/server-ts/access/jwt/index.js [hunk 2]
```diff
@@ -6,37 +6,49 @@ import resolvers from './resolvers';
 import schema from './schema.graphql';
 import AccessModule from '../AccessModule';
 import settings from '../../../../../settings';
+import { MESSAGE_APPEND_CONTEXT } from '../errorMessages';
 
-const grant = async user => {
-  const refreshSecret = settings.user.secret + user.passwordHash;
-  const [accessToken, refreshToken] = await createTokens(user, settings.user.secret, refreshSecret);
+const grant = async identity => {
+  const refreshSecret = settings.auth.secret + identity.passwordHash;
```
Comment (evgenijolshanskij) [inline modules/authentication/server-ts/access/jwt/index.js]: The Auth module doesn't know what fields `identity` consists of.

File: modules/authentication/server-ts/access/jwt/resolvers.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  Mutation: {
+    async refreshTokens(obj, { refreshToken: inputRefreshToken }, { getIdentity, getHash }) {
+      const decodedToken = jwt.decode(inputRefreshToken);
+      const isValidToken = !decodedToken || !decodedToken.id;
+
+      !isValidToken && throwError(MESSAGE_INVALID_REFRESH);
+      !getIdentity && throwError(MESSAGE_GET_IDENTIFY);
+
+      const identity = await getIdentity(decodedToken.id);
+      const refreshSecret = settings.auth.secret + getHash(decodedToken.id);
```
Comment (evgenijolshanskij) [inline modules/authentication/server-ts/access/jwt/resolvers.js]: How do you know the `getHash` function is exists?

File: modules/user/server-ts/social/facebook/index.js [hunk 4]
```diff
@@ -70,14 +70,16 @@ if (settings.user.auth.facebook.enabled && !__TEST__) {
       const user = await User.getUser(req.user.id);
       const redirectUrl = req.query.state;
       const tokens = await access.grantAccess(user, req);
+      const currentUser = await getCurrentUser(req, res);
```
Comment (evgenijolshanskij) [inline modules/user/server-ts/social/facebook/index.js]: Why did you add this line? There was not any `currentUser` data passed previously.

File: modules/user/server-ts/social/google/index.js [hunk 5]
```diff
@@ -1,11 +1,12 @@
 import { pick } from 'lodash';
 import passport from 'passport';
 import { OAuth2Strategy as GoogleStrategy } from 'passport-google-oauth';
+import { access } from '@module/authentication-server-ts';
```
Comment (evgenijolshanskij) [inline modules/user/server-ts/social/google/index.js]: Some middlewares should be located in `auth` module. These files are needed to be split as well.

File: modules/user/client-react/auth/facebook/containers/FacebookButton.native.jsx [hunk 6]
```diff
@@ -14,9 +14,9 @@ import {
   btnTextContainer,
   btnText
 } from '@module/look-client-react-native/styles';
+import access from '@module/authentication-client-react';
```
[the remark this answers is not in the document]
Comment (BaloPhilip) [reply modules/user/client-react/auth/facebook/containers/FacebookButton.native.jsx]: done

File: modules/user/client-react/auth/facebook/containers/FacebookButton.web.jsx [hunk 7]
```diff
@@ -3,8 +3,8 @@ import { withApollo } from 'react-apollo';
 import faFacebookSquare from '@fortawesome/fontawesome-free-brands/faFacebookSquare';
 import FontAwesomeIcon from '@fortawesome/react-fontawesome';
 import { Button } from '@module/look-client-react';
+import access from '@module/authentication-client-react';
```
[the remark this answers is not in the document]
Comment (BaloPhilip) [reply modules/user/client-react/auth/facebook/containers/FacebookButton.web.jsx]: done

File: modules/user/client-react/auth/github/containers/GitHubButton.native.jsx [hunk 8]
```diff
@@ -14,9 +14,9 @@ import {
   btnTextContainer,
   btnText
 } from '@module/look-client-react-native/styles';
+import access from '@module/authentication-client-react';
```
[the remark this answers is not in the document]
Comment (BaloPhilip) [reply modules/user/client-react/auth/github/containers/GitHubButton.native.jsx]: done

File: modules/user/client-react/auth/github/containers/GitHubButton.web.jsx [hunk 9]
```diff
@@ -3,8 +3,8 @@ import { withApollo } from 'react-apollo';
 import faGithubSquare from '@fortawesome/fontawesome-free-brands/faGithubSquare';
 import FontAwesomeIcon from '@fortawesome/react-fontawesome';
 import { Button } from '@module/look-client-react';
+import access from '@module/authentication-client-react';
```
[the remark this answers is not in the document]
Comment (BaloPhilip) [reply modules/user/client-react/auth/github/containers/GitHubButton.web.jsx]: done

File: modules/authentication/server-ts/access/jwt/createTokens.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import { MESSAGE_IDENTITY_WITHOUT_ID } from '../errorMessages';
+
+const { tokenExpiresIn, refreshTokenExpiresIn } = settings.auth.jwt;
+
+const createTokens = async (identity, secret, refreshSecret) => {
+  if (!identity.id) {
+    throw new AuthenticationError(MESSAGE_IDENTITY_WITHOUT_ID);
+  }
+
+  const tokenIdentity = pick(identity, ['id', 'username', 'role']);
```
[the remark this answers is not in the document]
Comment (basilalex) [reply modules/authentication/server-ts/access/jwt/createTokens.js]: Done

File: packages/server/src/modules.ts [hunk 11]
```diff
@@ -10,6 +10,7 @@ import upload from '@module/upload-server-ts';
 import subscription from '@module/payments-server-ts';
 import mailer from '@module/mailer-server-ts';
 import graphqlTypes from '@module/graphql-types-server-ts';
+import authenticate from '@module/authentication-server-ts';
```
[the remark this answers is not in the document]
Comment (basilalex) [reply packages/server/src/modules.ts]: Done.

File: modules/authentication/server-ts/access/jwt/index.js (hunk 2 above)
Comment (basilalex) [reply modules/authentication/server-ts/access/jwt/index.js]: Done.

File: modules/authentication/server-ts/access/jwt/resolvers.js (hunk 3 above)
Comment (basilalex) [reply modules/authentication/server-ts/access/jwt/resolvers.js]: Done.

File: modules/user/server-ts/social/facebook/index.js (hunk 4 above)
Comment (BaloPhilip) [reply modules/user/server-ts/social/facebook/index.js]: done

File: modules/user/server-ts/social/google/index.js (hunk 5 above)
Comment (basilalex) [reply modules/user/server-ts/social/google/index.js]: Done.

File: modules/authentication/client-react/access/jwt/components/Loading.jsx [hunk 12]
```diff
@@ -1,8 +1,10 @@
 import React from 'react';
 import { LayoutCenter } from '@module/look-client-react';
 
-export default (
+const Loading = () => (
   <LayoutCenter>
     <div className="text-center">App is loading...</div>
```
Comment (alekseyzadorozhniy) [inline modules/authentication/client-react/access/jwt/components/Loading.jsx]: Can we move this text to a file with localization?

File: modules/user/server-ts/social/password/index.js [hunk 13]
```diff
@@ -1,6 +1,6 @@
+import AuthModule from '@module/authentication-server-ts/social/AuthModule';
```
Comment (alekseyzadorozhniy) [inline modules/user/server-ts/social/password/index.js]: It is better to export/import `AuthModule` from `@module/authentication-server-ts` directly

File: modules/user/server-ts/testHelpers/index.js [hunk 14]
```diff
@@ -1,5 +1,5 @@
 import LOGIN from '@module/user-client-react/graphql/Login.graphql';
-import LOGOUT from '@module/user-client-react/access/session/graphql/Logout.graphql';
+import LOGOUT from '@module/authentication-client-react/access/session/graphql/Logout.graphql';
```
Comment (alekseyzadorozhniy) [inline modules/user/server-ts/testHelpers/index.js]: It is better to export/import `LOGOUT` and `LOGIN` from modules directly

File: modules/authentication/server-ts/access/session/index.js [hunk 15]
```diff
@@ -4,22 +4,17 @@ import { writeSession, createSession, readSession } from './sessions';
...
+  const session = { ...req.session, id };
   req.session = writeSession(req, session);
 };
 
-const getCurrentUser = async ({ req }) => {
-  if (req && req.session.userId) {
-    return await User.getUser(req.session.userId);
+const getCurrentIdentity = async ({ req, getIdentity }) => {
+  if (req && req.session.id) {
+    return await getIdentity(req.session.id);
```
Comment (evgenijolshanskij) [inline modules/authentication/server-ts/access/session/index.js]: There is no need to put `await` before `return`.

File: modules/authentication/server-ts/access/session/index.js [hunk 16]
```diff
@@ -32,30 +27,31 @@ const attachSession = req => {
...
+    throw new Error(MESSAGE_APPEND_CONTEXT);
+  }
+
   attachSession(req);
-  const user = context.user || (await getCurrentUser({ req, connectionParams, webSocket }));
-  const auth = {
-    isAuthenticated: !!user,
-    scope: user ? scopes[user.role] : null
-  };
+  const identity = context.identity || (await getCurrentIdentity({ req, getIdentity }));
```
Comment (evgenijolshanskij) [inline modules/authentication/server-ts/access/session/index.js]: `getIdentity` is needed to be checked.

File: modules/user/client-react/components/LoginView.native.jsx [hunk 17]
```diff
@@ -16,19 +16,30 @@ class LoginView extends React.PureComponent {
   }
 
   componentWillUnmount() {
-    Linking.removeListener('url');
+    Linking.removeEventListener('url', this.handleOpenURL);
   }
 
   handleOpenURL = async ({ url }) => {
+    if (!url.match(/data=([^#]+)/)) return;
```
Comment (evgenijolshanskij) [inline modules/user/client-react/components/LoginView.native.jsx]: Use constant for regexp to avoid duplications.

File: modules/authentication/client-react/access/jwt/components/Loading.jsx [hunk 18]
```diff
@@ -1,8 +1,10 @@
 import React from 'react';
 import { LayoutCenter } from '@module/look-client-react';
 
-export default (
+const Loading = () => (
```
Comment (evgenijolshanskij) [inline modules/authentication/client-react/access/jwt/components/Loading.jsx]: Why did you change the `export default` implementation? Is there some need to use this component as a `const`?

Comment (evgenijolshanskij) [reply modules/authentication/client-react/access/jwt/components/Loading.jsx]: And why the `App is loading...` component is located in the `auth` module?

File: modules/authentication/server-ts/access/session/index.js (hunk 16 above)
Comment (basilalex) [reply modules/authentication/server-ts/access/session/index.js]: Done.

File: modules/authentication/client-react/access/jwt/components/Loading.jsx (hunk 12 above)
Comment (BaloPhilip) [reply modules/authentication/client-react/access/jwt/components/Loading.jsx]: done

File: modules/user/client-react/components/LoginView.native.jsx (hunk 17 above)
Comment (BaloPhilip) [reply modules/user/client-react/components/LoginView.native.jsx]: done

File: modules/authentication/client-react/access/jwt/components/Loading.jsx (hunk 18 above)
Comment (BaloPhilip) [reply modules/authentication/client-react/access/jwt/components/Loading.jsx]: done

File: modules/authentication/client-react/access/index.js [hunk 19]
```diff
@@ -6,19 +6,21 @@ import session from './session';
...
 
+import resources from './locales';
+
 const ref = React.createRef();
 
-const clearApolloStoreAndReloadComponent = async client => {
-  await client.clearStore();
-  ref.current.reloadPage();
+const resetApolloCache = async client => {
+  await client.cache.reset();
```
Comment (vlasenko) [inline modules/authentication/client-react/access/index.js]: Why did we change approach of resetting the cache here? We were doing `client.clearStore()`, why did we change it?

File: modules/authentication/client-react/access/jwt/index.jsx [hunk 20]
```diff
@@ -4,15 +4,16 @@ import { withApollo } from 'react-apollo';
 import PropTypes from 'prop-types';
 
 import { getItem, setItem, removeItem } from '@module/core-common/clientStorage';
+import { CURRENT_USER_QUERY } from '@module/user-client-react';
```
Comment (vlasenko) [inline modules/authentication/client-react/access/jwt/index.jsx]: It contradicts our conception here, that auth module should be independent of concrete user module

File: modules/authentication/server-ts/access/jwt/resolvers.js [hunk 21]
```diff
@@ -0,0 +1,37 @@
+import jwt from 'jsonwebtoken';
+import { AuthenticationError } from 'apollo-server-errors';
+import createTokens from './createTokens';
+import settings from '../../../../../settings';
+import { MESSAGE_INVALID_REFRESH, MESSAGE_GET_IDENTIFY } from '../errorMessages';
+
+const throwError = message => {
```
Comment (vlasenko) [inline modules/authentication/server-ts/access/jwt/resolvers.js]: Avoid hiding throw, you win very little, but loose code readability

File: modules/authentication/server-ts/social/linkedIn/index.js [hunk 22]
```diff
@@ -0,0 +1,37 @@
+import passport from 'passport';
+import { Strategy as LinkedInStrategy } from 'passport-linkedin-oauth2';
+import settings from '../../../../../settings';
+import AuthModule from '../AuthModule';
+
+const { clientID, clientSecret, scope, callbackURL, enabled } = settings.auth.social.linkedin;
+
+const middleware = (app, { data }) => {
```
Comment (vlasenko) [inline modules/authentication/server-ts/social/linkedIn/index.js]: Lets rename `data` into `context` in modules, so that we have here `{ context }`

File: modules/authentication/client-react/access/index.js (hunk 19 above)
Comment (BaloPhilip) [reply modules/authentication/client-react/access/index.js]: When a user performs an authorization on the mobile app via social network, the user is being redirected to a third-party authentication service. After the auth completed the user is taken back to the app, but the application, by some reason, doesn't trigger re-update after the cache is cleaned using `client.clearStore()`. Replacing the `client.clearStore()` with `client.cache.reset()` does the trick. I can assume that `clearStore()` functions doesn't do well with subscriptions so that components are not aware about clearing the cahe.

Comment (vlasenko) [reply modules/authentication/client-react/access/index.js]: "by some reason" - is not a valid explanations. if you do not understand exactly what the problem is you cannot solve it. And your current solution is invalid, because you don't understand what is the exact problem

Comment (BaloPhilip) [reply modules/authentication/client-react/access/index.js]: In order to figure out why the `client.clearStore()` behaves different I have made a very deep research and step-by-step logging inside `apollo-client` library. Basically, the `client.clearStore()` invokes `cache.reset()` function under the hood and it should work the same way, but it doesn't. I made assumptions that the reason could in queries and migrations rejections that `client.clearStore()` function makes before `cache.reset()` invocation, so I disabled that functionality but still with no luck - the behavior left the same. Both the `client.clearStore()` and `cache.reset()` invoke the same function of the same object `netCache.reset()` of ApolloCacheRouter as a result. After the `client.clearStore()` being invoked components not re-render. I have inspected every line of cache clearing functionality and so far I don't have any idea what exactly can block subscription invocation for components.

Also, I tried to invoke `client.clearStore()` and `cache.reset()` one be one (in different order) and got the same result as I run `client.clearStore()` only.

File: modules/authentication/client-react/access/jwt/index.jsx (hunk 20 above)
Comment (BaloPhilip) [reply modules/authentication/client-react/access/jwt/index.jsx]: Done

File: modules/authentication/client-react/package.json [hunk 23]
```diff
@@ -0,0 +1,5 @@
+{
+  "name": "@gqlapp/authentication-client-react",
+  "version": "1.0.0",
```
Comment (vlasenko) [inline modules/authentication/client-react/package.json]: Please check what version is used in other modules nearby and do the same

Comment (BaloPhilip) [reply modules/authentication/client-react/package.json]: Done

File: config/user.js [hunk 24]
```diff
@@ -1,20 +1,6 @@
...
-      session: {
-        enabled: true
-      },
-      jwt: {
-        enabled: true,
-        tokenExpiresIn: '1m',
-        refreshTokenExpiresIn: '7d'
-      }
-    },
     password: {
```
Comment (vlasenko) [inline config/user.js]: Why do we leave auth configs for `user` and `certificate` authentication here?

File: modules/authentication/server-ts/access/errorMessages.js [hunk 25]
```diff
@@ -0,0 +1,9 @@
+export const MESSAGE_APPEND_CONTEXT = 'Can not find "appendContext" method. Please, add this method to the context.';
```
Comment (vlasenko) [inline modules/authentication/server-ts/access/errorMessages.js]: Perhaps we should use i18n here?

File: modules/authentication/server-ts/access/jwt/index.js [hunk 26]
```diff
@@ -6,37 +6,49 @@ import resolvers from './resolvers';
 import schema from './schema.graphql';
 import AccessModule from '../AccessModule';
 import settings from '../../../../../settings';
+import { MESSAGE_APPEND_CONTEXT } from '../errorMessages';
 
-const grant = async user => {
-  const refreshSecret = settings.user.secret + user.passwordHash;
-  const [accessToken, refreshToken] = await createTokens(user, settings.user.secret, refreshSecret);
+const grant = async (identity, req, hash = '') => {
```
Comment (vlasenko) [inline modules/authentication/server-ts/access/jwt/index.js]: We don't need `req` here, isn't it?

File: modules/module/server-ts/ServerModule.ts [hunk 27]
```diff
@@ -20,8 +20,10 @@ export interface ServerModuleShape extends CommonModuleShape {
   createResolversFunc?: Array<(pubsub: PubSub) => IResolvers>;
   createContextFunc?: Array<(props: CreateContextFuncProps) => { [key: string]: any }>;
   // Middleware
-  beforeware?: Array<(app: Express) => void>;
-  middleware?: Array<(app: Express) => void>;
+  beforeware?: Array<(app: Express, module: object) => void>;
```
Comment (vlasenko) [inline modules/module/server-ts/ServerModule.ts]: `module: object`,
I think it should be `modules`
and it should be `modules: ServerModuleShape`, or `modules: ServerModule`

File: modules/user/server-ts/social/password/resolvers.js [hunk 28]
```diff
@@ -2,17 +2,23 @@ import bcrypt from 'bcryptjs';
 import { pick, isEmpty } from 'lodash';
 import jwt from 'jsonwebtoken';
 import { UserInputError } from 'apollo-server-errors';
-import access from '../../access';
+import { access } from '@gqlapp/authentication-server-ts';
 import User from '../../sql';
 import settings from '../../../../../settings';
 
+const {
```
Comment (vlasenko) [inline modules/user/server-ts/social/password/resolvers.js]: I think this is not a good idea to make assignments like this on a module-level, it is okay to do them inside function, but not on a module-level. We loose code readability

File: modules/authentication/server-ts/access/errorMessages.js (hunk 25 above)
Comment (BaloPhilip) [reply modules/authentication/server-ts/access/errorMessages.js]: Done

File: modules/authentication/server-ts/access/jwt/index.js (hunk 26 above)
Comment (BaloPhilip) [reply modules/authentication/server-ts/access/jwt/index.js]: Done

File: config/user.js (hunk 24 above)
Comment (BaloPhilip) [reply config/user.js]: because all work with these settings occurs only in the user module

File: modules/module/server-ts/ServerModule.ts (hunk 27 above)
Comment (BaloPhilip) [reply modules/module/server-ts/ServerModule.ts]: Done

File: modules/authentication/client-react/access/index.js (hunk 19 above)
Comment (vlasenko) [reply modules/authentication/client-react/access/index.js]: The comment is incomplete. It should tell that we are calling this method from inside JWTLink in the middle of executing queries and we cannot cancel in-flight queries because of that

File: modules/user/server-ts/social/password/resolvers.js (hunk 28 above)
Comment (BaloPhilip) [reply modules/user/server-ts/social/password/resolvers.js]: Done

File: modules/authentication/client-react/access/index.js (hunk 19 above)
Comment (BaloPhilip) [reply modules/authentication/client-react/access/index.js]: Done

File: modules/counter/server-ts/index.ts [hunk 29]
```diff
@@ -7,5 +7,5 @@ import ServerModule from '@gqlapp/module-server-ts';
 export default new ServerModule({
   schema: [schema],
   createResolversFunc: [createResolvers],
-  createContextFunc: [() => ({ Counter: new Counter() })]
+  createGraphQLContextFunc: [() => ({ Counter: new Counter() })]
```
Comment (vlasenko) [inline modules/counter/server-ts/index.ts]: Theoretically this is a good change, but practically it is breaking change. You are forcing all existing projects based on the kit to do this renaming in their modules, when they upgrade. Is it worth it?

Comment (BaloPhilip) [reply modules/counter/server-ts/index.ts]: Done

**Comment by vlasenko** [comment]
Thank you @basilalex and @BaloPhilip ! Tons of work here!
